### PR TITLE
allow multiple competitive distances

### LINKED ISSUE #69: Allow for model selection between multiple competitive distances
Change [`run_cv()`](https://github.com/rudeboybert/forestecology/blob/master/R/model_functions.R#L256) to take vector argument of multiple `comp_dist` values and thus will involve different `focal_vs_comp` data frames

This should only be done for cross-validated RMSE values, as comparing non-cross-validated RMSE's based on overfit models is of little use.

### PR DESCRIPTION
**Pull request by simonpcouch**
Closes #69! 

``` r
library(forestecology)
library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union

run_cv(
  focal_vs_comp_ex,
  comp_dist = c(1, 2),
  blocks = blocks_ex
) %>%
  select(focal_ID, foldID, focal_sp, comp, comp_dist)
#> # A tibble: 12 x 5
#>    focal_ID foldID focal_sp       comp                 comp_dist
#>       <dbl> <fct>  <fct>          <list>                   <dbl>
#>  1        2 1      american_beech <tibble[,4] [2 × 4]>         1
#>  2        3 1      sugar_maple    <tibble[,4] [1 × 4]>         1
#>  3        4 1      american_beech <tibble[,4] [1 × 4]>         1
#>  4        5 1      sugar_maple    <tibble[,4] [1 × 4]>         1
#>  5        2 1      american_beech <tibble[,4] [2 × 4]>         2
#>  6        3 1      sugar_maple    <tibble[,4] [1 × 4]>         2
#>  7        4 1      american_beech <tibble[,4] [1 × 4]>         2
#>  8        5 1      sugar_maple    <tibble[,4] [1 × 4]>         2
#>  9        7 2      sugar_maple    <tibble[,4] [3 × 4]>         1
#> 10        9 2      sugar_maple    <tibble[,4] [3 × 4]>         1
#> 11        7 2      sugar_maple    <tibble[,4] [3 × 4]>         2
#> 12        9 2      sugar_maple    <tibble[,4] [3 × 4]>         2
```

<sup>Created on 2021-04-27 by the [reprex package](https://reprex.tidyverse.org) (v1.0.0.9001)</sup>

FYI, I haven't yet tested this on the `bw` examples or paper datasets. :-)

### CODE REVIEW

File: R/model_functions.R [hunk 1]
```diff
@@ -250,7 +250,7 @@ predict.comp_bayes_lm <- function(object, newdata, ...) {
 #'
 #' run_cv(
 #'   focal_vs_comp_ex,
-#'   comp_dist = 1,
+#'   comp_dist = c(1, 2),
 #'   blocks = blocks_ex
 #' )
 run_cv <- function(focal_vs_comp, comp_dist, blocks, prior_param = NULL, run_shuffle = FALSE) {
```
Comment (rudeboybert) [inline R/model_functions.R:256]: The `run_cv()` uses a static value of `comp_dist` to compute the `focal_vs_comp` data frame; rather it should change per each value of the `comp_dist` argument vector.

At first glance it seems that the function should be of rough form:
```r
run_cv_output <- 
  run_cv(
    growth_df = ,
    comp_dist = seq(from = 1, to = 15, by = 0.5),
    blocks = blocks_scbi,
    id = "stemID"
  )
```
Note the arguments match those of the `create_focal_vs_comp()` function so we can prolly do an `@inheritsParams` for the roxygen. 

Then internally `run_cv()` computes a new `focal_vs_comp` data frame for each value of `comp_dist`

**Review by rudeboybert — CHANGES_REQUESTED** [review]

**Comment by simonpcouch** [comment]
Ah, right! I'm on board.

I have one quick question about interface here that I'll leave for our meeting. :-)

**Comment by simonpcouch** [comment]
Okay, just pushed some changes!

Would appreciate a second set of eyes here on ensuring that the number of output rows is scaling as it should... I first thought this was a bug, but am feeling like it may be the right form now.

If this looks good, I'll go ahead and update documentation + examples!

``` r
library(forestecology)
library(tidyverse)

run_cv(
  growth_spatial_ex,
  comp_dist = 1,
  blocks = blocks_ex,
  id = "ID"
) %>%
  select(focal_ID, foldID, comp_dist, focal_sp, dbh, comp)
#> # A tibble: 6 x 6
#>   focal_ID foldID comp_dist focal_sp         dbh comp                
#>      <dbl> <fct>      <dbl> <fct>          <dbl> <list>              
#> 1        2 1              1 american_beech    20 <tibble[,4] [2 × 4]>
#> 2        3 1              1 sugar_maple       15 <tibble[,4] [1 × 4]>
#> 3        4 1              1 american_beech    12 <tibble[,4] [1 × 4]>
#> 4        5 1              1 sugar_maple       35 <tibble[,4] [1 × 4]>
#> 5        7 2              1 sugar_maple       22 <tibble[,4] [3 × 4]>
#> 6        9 2              1 sugar_maple       42 <tibble[,4] [3 × 4]>

run_cv(
  growth_spatial_ex,
  comp_dist = c(1, 2),
  blocks = blocks_ex,
  id = "ID"
) %>%
  select(focal_ID, foldID, comp_dist, focal_sp, dbh, comp)
#> # A tibble: 24 x 6
#>    focal_ID foldID comp_dist focal_sp         dbh comp                
#>       <dbl> <fct>      <dbl> <fct>          <dbl> <list>              
#>  1        2 1              1 american_beech    20 <tibble[,4] [2 × 4]>
#>  2        3 1              1 sugar_maple       15 <tibble[,4] [1 × 4]>
#>  3        4 1              1 american_beech    12 <tibble[,4] [1 × 4]>
#>  4        5 1              1 sugar_maple       35 <tibble[,4] [1 × 4]>
#>  5        7 2              1 sugar_maple       22 <tibble[,4] [3 × 4]>
#>  6        9 2              1 sugar_maple       42 <tibble[,4] [3 × 4]>
#>  7        2 1              1 american_beech    20 <tibble[,4] [4 × 4]>
#>  8        3 1              1 sugar_maple       15 <tibble[,4] [4 × 4]>
#>  9        4 1              1 american_beech    12 <tibble[,4] [3 × 4]>
#> 10        5 1              1 sugar_maple       35 <tibble[,4] [3 × 4]>
#> # … with 14 more rows

run_cv(
  growth_spatial_ex,
  comp_dist = c(1, 1.5, 2),
  blocks = blocks_ex,
  id = "ID"
) %>%
  select(focal_ID, foldID, comp_dist, focal_sp, dbh, comp)
#> # A tibble: 54 x 6
#>    focal_ID foldID comp_dist focal_sp         dbh comp                
#>       <dbl> <fct>      <dbl> <fct>          <dbl> <list>              
#>  1        2 1              1 american_beech    20 <tibble[,4] [2 × 4]>
#>  2        3 1              1 sugar_maple       15 <tibble[,4] [1 × 4]>
#>  3        4 1              1 american_beech    12 <tibble[,4] [1 × 4]>
#>  4        5 1              1 sugar_maple       35 <tibble[,4] [1 × 4]>
#>  5        7 2              1 sugar_maple       22 <tibble[,4] [3 × 4]>
#>  6        9 2              1 sugar_maple       42 <tibble[,4] [3 × 4]>
#>  7        2 1              1 american_beech    20 <tibble[,4] [2 × 4]>
#>  8        3 1              1 sugar_maple       15 <tibble[,4] [3 × 4]>
#>  9        4 1              1 american_beech    12 <tibble[,4] [2 × 4]>
#> 10        5 1              1 sugar_maple       35 <tibble[,4] [1 × 4]>
#> # … with 44 more rows
```

<sup>Created on 2021-05-10 by the [reprex package](https://reprex.tidyverse.org) (v1.0.0.9001)</sup>

**Comment by dallenmidd** [comment]
Hmm, that looks off. If you have *n* many focal trees and *d* many competitive distances, then the output should be approximately *nd*. It is approximately because the number of focal trees actually depends on the competitive distance, as the competitive distance increases more trees are included in the buffer. So actually I think this mulitple competitive distances actually needs to be implemented one step earlier, when we do `add_buffer_variable` to the `growth_df`. Or I guess `run_cv` with multiple distances needs to do that `add_buffer_variable` multiple times. 

But to answer the question, it looks like somethign funky is happening I would expect the first one to have 6 rows, then 12 (or just less than 12), and then 18 (or just less than 18).

**Comment by rudeboybert** [comment]
ooof, I think Dave is right about the buffer part. but if we make `run_cv()` for `add_buffer_variable()` on top of `create_focal_vs_comp()` and `fit_bayesian_model()` then basically we've just written a buffer for the entire algorithm! @simonpcouch could you hold off on further development until Dave and I have had a chance to discuss this?

**Comment by simonpcouch** [comment]
Closing re: Slack discussion!🐣

**Pull request closed without merging**